OrbitService/OrbitAsioServer.cpp

```diff
@@ -135,51 +135,59 @@ void OrbitAsioServer::TracingBufferThread() {
   while (tracing_handler_.IsStarted()) {
     Sleep(20);
 
-    std::vector<Timer> timers;
-    if (tracing_buffer_.ReadAllTimers(&timers)) {
-      Message Msg(Msg_Timers);
-      tcp_server_->Send(Msg, timers);
-    }
-
-    std::vector<LinuxCallstackEvent> callstacks;
-    if (tracing_buffer_.ReadAllCallstacks(&callstacks)) {
-      std::string message_data = SerializeObjectBinary(callstacks);
-      tcp_server_->Send(Msg_SamplingCallstacks, message_data.c_str(),
-                        message_data.size());
-    }
-
-    std::vector<CallstackEvent> hashed_callstacks;
-    if (tracing_buffer_.ReadAllHashedCallstacks(&hashed_callstacks)) {
-      std::string message_data = SerializeObjectBinary(hashed_callstacks);
-      tcp_server_->Send(Msg_SamplingHashedCallstacks, message_data.c_str(),
-                        message_data.size());
-    }
-
-    std::vector<ContextSwitch> context_switches;
-    if (tracing_buffer_.ReadAllContextSwitches(&context_switches)) {
-      Message Msg(Msg_ContextSwitches);
-      tcp_server_->Send(Msg, context_switches);
-    }
-
-    std::vector<LinuxAddressInfo> address_infos;
-    if (tracing_buffer_.ReadAllAddressInfos(&address_infos)) {
-      std::string message_data = SerializeObjectBinary(address_infos);
-      tcp_server_->Send(Msg_LinuxAddressInfos, message_data.c_str(),
-                        message_data.size());
-    }
-
-    std::vector<KeyAndString> keys_and_strings;
-    if (tracing_buffer_.ReadAllKeysAndStrings(&keys_and_strings)) {
-      std::string message_data = SerializeObjectBinary(keys_and_strings);
-      tcp_server_->Send(Msg_KeysAndStrings, message_data.c_str(),
-                        message_data.size());
-    }
-
-    std::vector<TidAndThreadName> tid_and_names;
-    if (tracing_buffer_.ReadAllThreadNames(&tid_and_names)) {
-      std::string message_data = SerializeObjectBinary(tid_and_names);
-      tcp_server_->Send(Msg_ThreadNames, message_data.c_str(),
-                        message_data.size());
-    }
+    SendBufferedMessages();
   }
+  // While sending the buffered messages, some buffers could again be filled.
+  // Send them now, so no message gets lost
+  SendBufferedMessages();
```

Comment:
Is it guaranteed that at this point, no thread is writing into the buffers anymore?

Reply:
I believe yes, but its a bit complex. So let me describe why I believe so.

We are leaving the while loop when, `LinuxTracingHandler::IsStarted()` evaluates to false, which is the case when `Tracer::IsTracing" is false (or the tracer itself is `nullptr`). This in turn is only false when either the `thread_` (which runs the `TracerThread` in its `Run` method) is nullptr or not `joinable`, thus not running. 
Now the Buffers only gets filled by the `LinuxTracingHandler::On...` functions, which again only gets called from within the `TracerThread`.
So after the while, no `TracerThread` is running, thus the buffers cannot be filled while sending the last couple of messages.

Reply:
And a new tracer thread can't yet have been started at this point, because OrbitAsioServer::StartCapture can't be called while we are still in StopCapture (assuming that's the only way that the stop is carried out). 

Ok, sounds correct to me.

Reply:
Right. Forgot to mention this.